colossalai/zero/sharded_optim/sharded_adam.py

```diff
@@ -0,0 +1,163 @@
+from enum import Enum
+from typing import Optional, Union
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from colossalai.amp.naive_amp._fp16_optimizer import DynamicGradScaler
+from colossalai.context.parallel_mode import ParallelMode
+from colossalai.core import global_context as gpc
+from colossalai.nn.optimizer import ColossalaiOptimizer
+from colossalai.zero.sharded_model import ShardedModelV2
+from torch import Tensor
+from torch.distributed import ProcessGroup
+from torch.optim import Optimizer
+
+from ._utils import has_inf_or_nan
+
+
+class OptimState(Enum):
+    SCALED = 1
+    UNSCALED = 2
+
+
+class ShardedAdam(ColossalaiOptimizer):
+
+    def __init__(self,
+                 adam_optim: Optimizer,
+                 sharded_model: nn.Module,
+                 cpu_offload: bool = False,
+                 initial_scale: float = 2**32,
+                 min_scale: float = 1,
+                 growth_factor: float = 2,
+                 backoff_factor: float = 0.5,
+                 growth_interval: float = 1000,
+                 hysteresis: float = 2,
+                 max_scale: int = 2**32,
+                 dp_process_group: Optional[ProcessGroup] = None,
+                 mp_process_group: Optional[ProcessGroup] = None) -> None:
+        super().__init__(adam_optim)
+        self.model: Union[nn.Module, ShardedModelV2] = sharded_model
+        self.model_is_sharded = isinstance(sharded_model, ShardedModelV2)
+        self.state_device = torch.cuda.current_device() if not cpu_offload else torch.device('cpu')
+        self.optim_state: OptimState = OptimState.UNSCALED
+        self.dp_process_group = dp_process_group or gpc.get_group(ParallelMode.DATA)
+        self.mp_process_group = mp_process_group or gpc.get_group(ParallelMode.MODEL)
+        # Grad scaler
+        self.grad_scaler = DynamicGradScaler(initial_scale=initial_scale,
+                                             min_scale=min_scale,
+                                             growth_factor=growth_factor,
+                                             backoff_factor=backoff_factor,
+                                             growth_interval=growth_interval,
+                                             hysteresis=hysteresis,
+                                             max_scale=max_scale)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.state_device)
+
+        # Early state initialization
```

Comment:
@feifeibear I think the sharded Adam is easy to expand toa  general sharded optimizer by just removing early state initialization here. And in step(), we have to make sure param.data has correct shape, dtype and device.

Reply:
I agree with you on moving OS memory allocation and init here.

Reply:
Agree. This part is not necessary as many torch optimizers will do lazy init.